loadbalancer/algorithm.go

```diff
@@ -89,19 +96,70 @@ func (*consistentHash) Apply(ctx *routing.LBContext) routing.LBEndpoint {
 	key := net.RemoteHost(ctx.Request).String()
 	if _, err := h.Write([]byte(key)); err != nil {
 		log.Errorf("Failed to write '%s' into hash: %v", key, err)
-		return ctx.Route.LBEndpoints[rand.Intn(len(ctx.Route.LBEndpoints))]
+		return ctx.Route.LBEndpoints.At(rand.Intn(ctx.Route.LBEndpoints.Length()))
 	}
 	sum = h.Sum32()
-	choice := int(sum) % len(ctx.Route.LBEndpoints)
+	choice := int(sum) % ctx.Route.LBEndpoints.Length()
 	if choice < 0 {
-		choice = len(ctx.Route.LBEndpoints) + choice
+		choice = ctx.Route.LBEndpoints.Length() + choice
+	}
+	return ctx.Route.LBEndpoints.At(choice)
+}
+
+type powerOfChoices struct{
+	rand *rand.Rand
+	numberOfChoices int
+	mutex sync.Mutex
+}
+
+// Selects N random backends and picks the one with less outstanding requests.
+func newPowerOfChoices(endpoints []string) routing.LBAlgorithm {
+	t := time.Now().UnixNano()
+	return &powerOfChoices{
+		rand: rand.New(rand.NewSource(t)),
+		numberOfChoices: 2, // TODO: make this the value part of skipper configuration.
+	}
+}
+func (a *powerOfChoices) GetRandomChoice(length int) int {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+	return a.rand.Intn(length)
+}
+
+func contains(s []int, e int) bool {
+	for _, a := range s {
+		if a == e {
+			return true
+		}
 	}
-	return ctx.Route.LBEndpoints[choice]
+	return false
+}
+
+func (a *powerOfChoices) Apply(ctx *routing.LBContext) routing.LBEndpoint {
+	chosen := make([]int, 0, a.numberOfChoices)
+	for i := 0 ; i < a.numberOfChoices ; i++  {
+		candidate := a.GetRandomChoice(ctx.Route.LBEndpoints.Length())
+		// Pick a different endpoint if was already selected
+		for contains(chosen, candidate) {
+			candidate = a.GetRandomChoice(ctx.Route.LBEndpoints.Length())
+		}
+		chosen = append(chosen, candidate)
+	}
+	bestEndpoint := ctx.Route.LBEndpoints.At(chosen[0])
+	bestScore := bestEndpoint.Metrics.GetInflightRequests()
```

Comment:
These specific calls we can replace with our stored func in our struct.

Reply:
Done.